### ISSUE TITLE
Action required: Let's Encrypt subscriber agreement URL Change

### ISSUE THREAD

**Issue by cpu** [webmin/webmin#699]
Hi there,

Your project has a hardcoded URL reference to the [current](https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf) Let's Encrypt subscriber agreement:

https://github.com/webmin/webmin/blob/1cf63f8249f6c1ae1b232dfecca597b2af18fb95/webmin/acme_tiny.py#L99

This URL [**changed November 15th**](https://community.letsencrypt.org/t/subscriber-agreement-update-november-15-2017/45607). New registration requests using this ACME client will produce an error as of today.

One Webmin user has already reported this to the Certbot repo by mistake: https://github.com/certbot/certbot/issues/5240

ACME clients do not need to hardcode a subscriber agreement URL and Let's Encrypt discourages this practice. You should learn the current agreement URL at runtime instead so that it is always current.

One option is to reference the agreement URL from the `"meta"` key's `"terms-of-service"` element from the response to a `GET` request to the ACME server's `/directory` endpoint:
```
{
  <snip>
  "meta": {
        "terms-of-service": "https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf"
  },
  <snip>
}
```

A second option is to learn the current subscriber agreement URL at runtime as part of the `new-reg` flow: Client code can submit an initial `new-reg` request without an `agreement` value in the request payload. The account will be created and in the response will be a `Link` header with a `terms-of-service` relation pointing at the current agreement URL. Once this `Link` header has been seen the client should update the registration to agree to the terms by sending a payload with the `agreement` value set to the current agreement URL.

Thanks!

**Comment by jcameron** [comment]
Thanks, we'll fix this.

Silly question though - what value does a subscriber agreement have it if is never actually seen by a human and acknowledged entirely in software?

**Comment by chris001** [comment]
If I may interject!  The automated software client (webmin) should take the link to the current agreement, received from LE during this communications flow, and display the page contents referred to by the link (the user agreement/terms of service) to the human user of wemin. Explaining to them, this is the LE TOS, by using LE, you are agreeing to these terms, click to Agree / Cancel.

**Comment by jcameron** [comment]
That would actually be reasonable! Although I'd prefer not to do it, as users would likely prefer that the Let's Encrypt function just work with minimal human intervention.

**Comment by chris001** [comment]
True. Could probably just say, By using LE SSL/TLS certs, you agree to the LE terms. Click here to read them in case you want to read it. When they click, webmin wget's the LE T and C page html and displays it.

**Comment by cpu** [comment]
@jcameron I understand your perspective, but is it any different from having a ToS URL hardcoded in source code that users won't read? 

I think a message wherever the user enables the Let's Encrypt plugin that mentions doing so means agreeing to the subscriber agreement (with a link) would be sufficient if you don't want to interactively prompt per @chris001's suggestion.

**Comment by nigel-aves** [comment]
@jcameron I'm the user who inadvertently reported the "error" to Certbot.

As a user I think that @cpu has the best solution, I also think it satisfies Jamie's idea about making the process work with little user intervention. If the wording was changed from "Request Certificate" to "Agree ToS and Request Certificate", and below that have the link to the ToS, then it's up to the user. It would also mean that if they are trying to set up a large number of web sites to use Certbot they would only have to read once (if they even read it all :) ) .

The ToS is 7 pages long. That's rather a lot to have to keep scrolling through to find the Accept button at the bottom of the page.

Nigel.

**Comment by chris001** [comment]
True. The LE ToS and its (passive, implicit) acceptance should be made visible in the Webmin UI.  Similar to websites which make the cookie disclaimer js toaster alert bar, to comply with EU privacy law.  By using LE certs in Webmin, they implicitly agree to the LE ToS.  Webmin should provide the dynamically obtained link to the full LE ToS in its UI, for those who care to read it.

**Comment by icloudiap** [comment]
line 99 replace -> <strong>"agreement": "https://letsencrypt.org/documents/LE-SA-v1.1.1-August-1-2016.pdf", 
with -> "agreement": json.loads(urlopen(CA + "/directory").read().decode('utf8'))['meta']['terms-of-service'],
hope this helpful
source : https://github.com/diafygi/acme-tiny/blob/master/acme_tiny.py#L85

**Comment by jcameron** [comment]
I like the idea of making the agreement visible in the UI.

**Comment by inductivesoul** [comment]
As long as we get a timely and proper rework of the LE code I will be happy with many of the above suggestions, as a user of this service on Webmin, I can say that this is needing to be reprogrammed so it does not break in the future and would automatically grab the new TOS/TOU agreements. Its important to have the option to read them and or just click install.

**Comment by iliaross** [comment]
❗️ If the issue isn't resolved, feel free to reopen it with updated details.
